src/node_modules.cc

```diff
@@ -0,0 +1,442 @@
+using v8::FunctionCallbackInfo;
+using v8::Global;
+using v8::HandleScope;
+using v8::Isolate;
+using v8::JSON;
+using v8::Local;
+using v8::Map;
+using v8::MaybeLocal;
+using v8::NewStringType;
+using v8::Object;
+using v8::ObjectTemplate;
+using v8::Primitive;
+using v8::String;
+using v8::Undefined;
+using v8::Value;
+
+#ifdef __POSIX__
+constexpr char kPathSeparator = '/';
+#else
+const char* const kPathSeparator = "\\/";
+#endif
+
+void BindingData::MemoryInfo(MemoryTracker* tracker) const {
+  // Do nothing
+}
+
+BindingData::BindingData(Realm* realm,
+                         v8::Local<v8::Object> object,
+                         InternalFieldInfo* info)
+    : SnapshotableObject(realm, object, type_int), package_configs_() {}
+
+bool BindingData::PrepareForSerialization(v8::Local<v8::Context> context,
+                                          v8::SnapshotCreator* creator) {
+  // Return true because we need to maintain the reference to the binding from
+  // JS land.
+  return true;
+}
+
+InternalFieldInfoBase* BindingData::Serialize(int index) {
+  DCHECK_IS_SNAPSHOT_SLOT(index);
+  InternalFieldInfo* info =
+      InternalFieldInfoBase::New<InternalFieldInfo>(type());
+  return info;
+}
+
+void BindingData::Deserialize(v8::Local<v8::Context> context,
+                              v8::Local<v8::Object> holder,
+                              int index,
+                              InternalFieldInfoBase* info) {
+  DCHECK_IS_SNAPSHOT_SLOT(index);
+  v8::HandleScope scope(context->GetIsolate());
+  Realm* realm = Realm::GetCurrent(context);
+  BindingData* binding = realm->AddBindingData<BindingData>(holder);
+  CHECK_NOT_NULL(binding);
+}
+
+// TODO(@anonrig): Add actual package.json file string to output
+// depending on JS equivalent of:
+// manifest = getOptionValue('--experimental-policy') ?
+// require('internal/process/policy').manifest : null;
+Local<Array> BindingData::PackageConfig::Serialize(Realm* realm) {
+  auto isolate = realm->isolate();
+  const auto ToString = [isolate](std::string_view input) -> Local<Primitive> {
+    return String::NewFromUtf8(
+               isolate, input.data(), NewStringType::kNormal, input.size())
+        .ToLocalChecked();
+  };
+  Local<Value> values[5] = {
+      name.has_value() ? ToString(*name) : Undefined(isolate),
+      main.has_value() ? ToString(*main) : Undefined(isolate),
+      ToString(type),
+      imports.has_value() ? ToString(*imports) : Undefined(isolate),
+      exports.has_value() ? ToString(*exports) : Undefined(isolate)};
+  return Array::New(isolate, values, 5);
+}
+
+std::optional<BindingData::PackageConfig> BindingData::GetPackageJSON(
+    Realm* realm, std::string_view path, ErrorContext* error_context) {
+  auto binding_data = realm->GetBindingData<BindingData>();
+  auto package_configs_ = binding_data->package_configs_;
+
+  auto cache_entry = package_configs_.find(std::string(path));
+  if (cache_entry != package_configs_.end()) {
```

Comment:
Should this re-check read permissions before returning the cached entry?

Reply:
Afaik permissions does not change after Node is booted. But lets ask the author of permission model: @RafaelGSS?